components/core/src/clp/ffi/ir_stream/Serializer.cpp

```diff
@@ -59,6 +258,207 @@ auto Serializer<encoded_variable_t>::change_utc_offset(UtcOffset utc_offset) ->
+            }
+        } else {
+            // Serialize primitive values
+            if (false
+                == (serialize_key(schema_tree_node_id) && serialize_val(val, schema_tree_node_type)
+                ))
+            {
+                failure = true;
+                break;
+            }
+        }
+    }
+
+    if (failure) {
+        m_schema_tree.revert();
+        return false;
+    }
+
+    m_ir_buf.insert(
+            m_ir_buf.cend(),
+            m_schema_tree_node_buf.cbegin(),
+            m_schema_tree_node_buf.cend()
+    );
+    m_ir_buf.insert(m_ir_buf.cend(), m_key_group_buf.cbegin(), m_key_group_buf.cend());
+    m_ir_buf.insert(m_ir_buf.cend(), m_value_group_buf.cbegin(), m_value_group_buf.cend());
+    return true;
+}
+
+template <typename encoded_variable_t>
+auto Serializer<encoded_variable_t>::serialize_schema_tree_node(
+        SchemaTree::NodeLocator const& locator
+) -> bool {
+    switch (locator.get_type()) {
+        case SchemaTreeNode::Type::Int:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeInt);
+            break;
+        case SchemaTreeNode::Type::Float:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeFloat);
+            break;
+        case SchemaTreeNode::Type::Bool:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeBool);
+            break;
+        case SchemaTreeNode::Type::Str:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeStr);
+            break;
+        case SchemaTreeNode::Type::UnstructuredArray:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeUnstructuredArray);
+            break;
+        case SchemaTreeNode::Type::Obj:
+            m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeObj);
+            break;
+        default:
+            // Unknown type
+            return false;
+    }
+    auto const parent_id{locator.get_parent_id()};
+    if (parent_id <= UINT8_MAX) {
+        m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeParentIdUByte);
+        m_schema_tree_node_buf.push_back(static_cast<int8_t>(static_cast<uint8_t>(parent_id)));
+    } else if (parent_id <= UINT16_MAX) {
+        m_schema_tree_node_buf.push_back(cProtocol::Payload::SchemaTreeNodeParentIdUShort);
+        serialize_int(static_cast<uint16_t>(parent_id), m_schema_tree_node_buf);
+    } else {
+        // Out of range
+        return false;
+    }
+    return serialize_string_packet(locator.get_key_name(), m_schema_tree_node_buf);
+}
+
+template <typename encoded_variable_t>
+auto Serializer<encoded_variable_t>::serialize_key(SchemaTreeNode::id_t id) -> bool {
+    if (id <= UINT8_MAX) {
+        m_key_group_buf.push_back(cProtocol::Payload::KeyIdUByte);
+        m_key_group_buf.push_back(static_cast<int8_t>(static_cast<uint8_t>(id)));
```

Comment:
```suggestion
        m_key_group_buf.push_back(bit_cast<int8_t>(static_cast<uint8_t>(id)));
```

`bit_cast` is safer to ensure we're not unintentionally reducing the size of the type.